src/bin/mdbook-xgettext.rs

```diff
@@ -81,21 +90,242 @@ fn create_catalog(ctx: &RenderContext) -> anyhow::Result<Catalog> {
                 let source = format!("{}:{}", path.display(), lineno);
-                add_message(&mut catalog, &msgid, &source);
+                add_message(catalog, &msgid, &source);
             }
+
+            let file_path = if depth == 0 {
+                Some(default_file_path.display().to_string())
+            } else if depth == 1 {
+                Some(chapter.path.clone().unwrap().display().to_string())
+            } else {
+                None
+            };
+            get_subcontent_for_chapter(chapter.clone(), file_path, depth)
+                .iter()
+                .for_each(
+                    |ChapterTemplateInfo {
+                         content,
+                         source,
+                         template_destination,
+                     }| {
+                        extract_messages(content)
+                            .iter()
+                            .for_each(|(lineno, msgid)| {
+                                let path = if depth == 0 {
+                                    template_destination.to_owned()
+                                } else {
+                                    ctx.config
+                                        .book
+                                        .src
+                                        .join(template_destination)
+                                        .display()
+                                        .to_string()
+                                };
+                                if let None = catalogs.get(&path) {
+                                    catalogs.insert(
+                                        path.to_string(),
+                                        Catalog::new(generate_catalog_metadata(&ctx)),
+                                    );
+                                }
+                                let catalog = catalogs.get_mut(&path).unwrap();
+
+                                let source2 = format!(
+                                    "{}:{}",
+                                    ctx.config.book.src.join(source).display(),
+                                    lineno
+                                );
+                                add_message(catalog, &msgid, &source2);
+                            });
+                    },
+                );
         }
     }
 
-    Ok(catalog)
+    Ok(catalogs)
+}
+
+// Contains information necessary to create polib messages out
+// of chapter content and save them into a template file.
+#[derive(Clone, Debug, Hash)]
+struct ChapterTemplateInfo {
+    // The chapter's content.
+    content: String,
+    // The file where the content is sourced.
+    source: String,
+    // The output destination for the messages
+    // created with the content and source.
+    template_destination: String,
+}
+
+// Get ChapterTemplateInfo for sub-chapters, depending on the specified depth.
+fn get_subcontent_for_chapter(
+    c: Chapter,
```

Comment:
From the name, I would expect `get_subcontent_for_chapter` to be a read-only operation on a `Chapter`. So this should most likely be `chapter: &Chapter` to show that we only borrow the chapter in a non-destructive sense. This would immediately allow you to avoid deep-cloning the chapter at the call site above.